### PR TITLE
fix(kuma-dp): Fix envoy binary not found 

### LINKED ISSUE #686: When `envoy` is not found `kuma-dp` terminates in an inconsistent state
When the `envoy` executable is not found, `kuma-dp` terminates in an inconsistent state:

```shell
$ kuma-dp run   --name=dp-echo-1   --mesh=default   --cp-address=http://127.0.0.1:5681   --dataplane-token-file=/tmp/kuma-dp-echo-1
Error: could not find binary in any of the following paths: [envoy /Users/marco/git/kuma/build/artifacts-darwin-amd64/kuma-dp/envoy /Users/marco/git/kuma/envoy]
$ kuma-dp run   --name=dp-echo-1   --mesh=default   --cp-address=http://127.0.0.1:5681   --dataplane-token-file=/tmp/kuma-dp-echo-1
Error: listen unix /tmp/kuma-access-logs-dp-echo-1-default.sock: bind: address already in use
```

It seems like the file `kuma-access-logs-dp-echo-1-default.sock` is being always created even if `kuma-dp` cannot start.

### PR DESCRIPTION
**Pull request by tharun208**
Check for envoy path and throw an error if not found during initializing envoy
Updated tests

### Fix Envoy Binary Not Found

When envoy is not found Kuma-dp stops.

### Full changelog

* Checked for envoy path and if present creates an envoy instance.
* Updated tests.

### Issues resolved

Fix #686

### CODE REVIEW

**Comment by tharun208** [comment]
> 1. could not find the "envoy" executable in your pat



> We need to implement two checks here:
> 
> 1. One to determine if `envoy` exists. If not then return an error that says something like `could not find the "envoy" executable in your path`
> 2. One to return an error if running envoy fails for whatever reason (we have this here in this PR)
> 3. And we could even add a check to determine that the version of `envoy` that we found (`envoy --version`) is compatible with the current version of Kuma.
> 
> For the last step, we will need to have a list of version ranges that we support (ie from `1.12` to `1.14`) so that we can support minor releases that are being created in the meanwhile by Envoy without having to update Kuma. I will let @jakubdyszkiewicz chime in for this.

1. I added the envoy exists check and the error message.
2. Running envoy fails is already present.

**Comment by tharun208** [comment]
> > 1. could not find the "envoy" executable in your pat
> 
> > We need to implement two checks here:
> > 
> > 1. One to determine if `envoy` exists. If not then return an error that says something like `could not find the "envoy" executable in your path`
> > 2. One to return an error if running envoy fails for whatever reason (we have this here in this PR)
> > 3. And we could even add a check to determine that the version of `envoy` that we found (`envoy --version`) is compatible with the current version of Kuma.
> > 
> > For the last step, we will need to have a list of version ranges that we support (ie from `1.12` to `1.14`) so that we can support minor releases that are being created in the meanwhile by Envoy without having to update Kuma. I will let @jakubdyszkiewicz chime in for this.
> 

Hi Marco,
> 1. I added the envoy exists check and the error message.
> 2. Running envoy fail check is already present.

**Comment by jakubdyszkiewicz** [comment]
>And we could even add a check to determine that the version of envoy that we found (envoy --version) is compatible with the current version of Kuma.
>
> For the last step, we will need to have a list of version ranges that we support (ie from 1.12 to 1.14) so that we can support minor releases that are being created in the meanwhile by Envoy without having to update Kuma. I will let @jakubdyszkiewicz chime in for this.


I am hesitant about this step @subnetmarco 
It is very much a use case to use your custom Envoy distro with Kuma. The version name then could be anything. Additionally, I'd allow users to experiment with new versions of Envoy, especially if this is a minor security release.

If we want to do it, we should only validate major version so in case of `1.12.1` we let `1.12.*`. Additionally, there should be a flag to skip this validation.

**Comment by tharun208** [comment]
> > And we could even add a check to determine that the version of envoy that we found (envoy --version) is compatible with the current version of Kuma.
> > For the last step, we will need to have a list of version ranges that we support (ie from 1.12 to 1.14) so that we can support minor releases that are being created in the meanwhile by Envoy without having to update Kuma. I will let @jakubdyszkiewicz chime in for this.
> 
> I am hesitant about this step @subnetmarco
> It is very much a use case to use your custom Envoy distro with Kuma. The version name then could be anything. Additionally, I'd allow users to experiment with new versions of Envoy, especially if this is a minor security release.
> 
> If we want to do it, we should only validate major version so in case of `1.12.1` we let `1.12.*`. Additionally, there should be a flag to skip this validation.

@subnetmarco are we going to add this check?

File: app/kuma-dp/cmd/run.go [hunk 1]
```diff
@@ -100,13 +100,17 @@ func newRunCmd() *cobra.Command {
...
-			dataplane := envoy.New(envoy.Opts{
+			dataplane, err := envoy.New(envoy.Opts{
 				Catalog:   catalog,
 				Config:    cfg,
 				Generator: bootstrapGenerator,
 				Stdout:    cmd.OutOrStdout(),
 				Stderr:    cmd.OutOrStderr(),
 			})
+			if err != nil {
+				runLog.Error(err, `could not find the "envoy" executable in your path`)
```
Comment (subnetmarco) [inline app/kuma-dp/cmd/run.go:111]: What happens if `envoy` is found but running the executable fails for some reasons (ie, corrupted Envoy executable)? I would like to implement two separate checks:

* First determine if `envoy` exists as a file before executing it, if not return `could not find the "envoy" executable in your path`.
* Then, if found, when executing `envoy` if it fails for some reason return a generic error `the "envoy" executable was found at "/path" but an error occurred when executing it`.

**Comment by subnetmarco** [comment]
Okay then we can skip the Envoy version range validation for now, but I added another comment to my review that needs to be addressed.

File: app/kuma-dp/pkg/dataplane/envoy/envoy_test.go [hunk 2]
```diff
@@ -174,5 +174,32 @@ var _ = Describe("Envoy", func() {
...
+			By("starting a mock dataplane")
+			// when
+			_, err := New(Opts{
+				Config:    cfg,
+				Generator: sampleConfig,
+				Stdout:    &bytes.Buffer{},
+				Stderr:    &bytes.Buffer{},
+			})
+			// then
+			Expect(err).NotTo(BeNil())
```
Comment (lobkovilya) [inline app/kuma-dp/pkg/dataplane/envoy/envoy_test.go:199]: Can we please also check the exact value of the error?

File: app/kuma-dp/pkg/dataplane/envoy/envoy.go [hunk 3]
```diff
@@ -38,8 +38,14 @@ type Opts struct {
 	Stderr    io.Writer
 }
 
-func New(opts Opts) *Envoy {
-	return &Envoy{opts: opts}
+func New(opts Opts) (*Envoy, error) {
+	binaryPathConfig := opts.Config.DataplaneRuntime.BinaryPath
+	_, err := lookupEnvoyPath(binaryPathConfig)
+	if err != nil {
```
Comment (lobkovilya) [inline app/kuma-dp/pkg/dataplane/envoy/envoy.go:44]: [minor]: I'd prefer if this one be one-lined:
```go
if _, err := lookupEnvoyPath(opts.Config.DataplaneRuntime.BinaryPath); err != nil {
    runLog.Error(err, "could not find the envoy executable in your path")
    return nil, err
}

Comment (tharun208) [reply app/kuma-dp/pkg/dataplane/envoy/envoy.go]: changed

**Pull request merged**